Issue 411  padding

kohr-h opened the pull request:
Implements the resizing of a multi-dimensional array as an operator wrapping a helper function `resize_array` which will be very helpful in itself.

Basically, resizing means keeping the cell sizes and changing the number of cells per axis, plus a shift by a multiple of the cell size.

The extension (padding) is supported in 5 different versions: constant (i.e. fill with a constant), periodic, symmetric, order 0 (continue constantly) and order 1 (continue with constant slope). Not sure we need all of those, but was not too hard to add once the infrastructure was there.

~~What's missing currently is the adjoint. It's not super simple and needs some thought to get it right. I'd suggest we deal with it in a separate issue.~~

TODO:
- [x] Refactor code so forward and adjoint use as much common code as possible (basically the slice calculations)
- [x] More tests, especially corner cases and adjoint
- [x] Doc improvements
- [x] Decision on how to specify start position / distribution of cropped stuff between the left and right sides of the array. I went for `offset` ~~Currently implemented:~~
  * ~~`num_left` -- how many cells added/removed on the left side~~
  * ~~`frac_left` -- same as above, but as fraction. Less reliable due to rounding. Will be removed.~~
  
   ~~Questions: Do we need a `num_right` counterpart? Do we need another option to specify the amount?~~

Edit: last point updated.

kohr-h commented on odl/discr/discr_ops.py:
```diff
@@ -151,6 +159,226 @@ def adjoint(self):
+        if self.pad_mode == 'constant' and self.pad_const != 0:
+            return ResizingOperator(
+                domain=self.domain, range=self.range, pad_mode='constant',
+                pad_const=0.0)
```
`else` missing

kohr-h replied:
added

adler-j replied:
Seem to miss several arguments here, like `discr_kwargs` and, `offset` and `ran_shape`.

kohr-h replied:
Forgot to update, thx.

kohr-h replied:
Actually no, these parameters are all encoded in `resize_op.range` so we don't need them.

kohr-h commented on odl/discr/discr_ops.py:
```diff
@@ -151,6 +159,226 @@ def adjoint(self):
+                                      'thus has no adjoint')
+        # TODO: this is not correct, fix it
+        return ResizingOperator(domain=self.range, range=self.domain,
+                                pad_mode='constant', pad_const=0.0)
```
Change to `NotImplementedError` (text also wrong)

kohr-h commented on odl/discr/discr_ops.py:
```diff
@@ -151,6 +159,226 @@ def adjoint(self):
+        num_left = np.around(num_left_float).astype(int)
+        if not np.allclose(num_left, num_left_float):
+            raise ValueError('range is shifted relative to domain by a '
+                             'non-multiple of cell_sides')
```
perhaps print shift

kohr-h commented on odl/discr/discr_ops.py:
```diff
@@ -151,6 +159,226 @@ def adjoint(self):
+            newshp, num_left):
+
+        n_diff = n_new - n_orig
+#        print(n_diff)
```
Remove commented-out print statements

kohr-h commented on odl/util/numerics.py:
```diff
@@ -297,6 +303,521 @@ def fast_1d_tensor_mult(ndarr, onedim_arrs, axes=None, out=None):
+        number of entries added to the left is gained by rounding to the
+        nearest integer. The default ``None`` is equivalent to 0.5.
+        This option can be specified per axis with a sequence.
+        Cannot be combined with ``num_left``.
```
Actually, if I think about it, this option should be removed. It's not super reliable due to rounding, and the same effect can easily be reached by a simple calculation by the user. Opinions?

adler-j replied:
Well the sheer length of the doc that is needed to explain it is a strong argument against keeping it.

kohr-h commented on odl/util/numerics.py:
```diff
@@ -297,6 +303,521 @@ def fast_1d_tensor_mult(ndarr, onedim_arrs, axes=None, out=None):
+
+        Note that for ``'symmetric'`` and ``'periodic'`` padding, the
+        number of added values on each side of the array cannot exceed
+        the original size.
```
Remove this paragraph

kohr-h commented on odl/util/numerics.py:
```diff
@@ -297,6 +303,521 @@ def fast_1d_tensor_mult(ndarr, onedim_arrs, axes=None, out=None):
+    Examples
+    --------
+    The input can be shrunk by simply providing a smaller size.
+    The left/right distribution of removed entries is ananogous to
```
analogous

kohr-h commented on odl/util/numerics.py:
```diff
@@ -297,6 +303,521 @@ def fast_1d_tensor_mult(ndarr, onedim_arrs, axes=None, out=None):
+
+
+def _resize_array_adj(arr, out, num_left, pad_mode='constant', pad_const=0):
+    """Adjoint operation to `resize_array`.
```
Shouldn't be part of this PR, remove the function.

kohr-h replied:
now part of the PR anyway

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -151,6 +159,226 @@ def adjoint(self):
+        return self.__pad_mode
+
+    @property
+    def pad_const(self):
```
I dont think this is the default name. Looking at [`scipy`](http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.ndimage.filters.gaussian_filter.html) they seem to go for `cval`. We may or may not want to go fro this.

kohr-h replied:
I chose `pad_const` since it makes the connection to `'constant'` padding obvious.
We use `padding_value` in [other places](https://github.com/odlgroup/odl/blob/281d1d8763cc33e62a0e67268a5668fa033bdd95/odl/discr/diff_ops.py#L55-L56), which I think is inferior. `cval` is only shorter, not better. Not sure if one can speak of a "default" here. Even the names for the padding modes are not really standardized. For example, the mode `'constant'` in PyWavelets is the "replicate boundary values" mode, not "fill constant" as in SciPy. So I'd say we're free to choose whichever name we think is most appropriate.

kohr-h replied:
ok?

adler-j replied:
Fine with this, but we should make an issue to standardize these across ODL.

kohr-h replied:
See #511

adler-j commented on doc/source/math/resizing_ops.rst:
```diff
@@ -0,0 +1,356 @@
+
+Introduction
+============
+Resizing can be described in a continuous as well as in a fully discrete setting. In the
```
This paragraph does not (at least to me) fully describe what you mean by resizing. "Changing the domain" is abit vague. Change in what way? It would be good to be abit more clear here since this is the most important paragraph.

kohr-h replied:
Will improve.

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -151,6 +159,226 @@ def adjoint(self):
+
+class ResizingOperator(Operator):
+
+    """Operator mapping between discretized spaces of different shapes.
```
Too vague

kohr-h replied:
Hard to be more precise in one line, but will try..

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -151,6 +159,226 @@ def adjoint(self):
+
+    """Operator mapping between discretized spaces of different shapes.
+
+    This operator is intended as a mapping between uniformly
```
"is intended as" is abit unnecessary, unless this is an abstract operator intended for inheritance.

kohr-h commented:
Some of the texts are not ideal yet, I'll get to that.

adler-j commented on doc/source/math/resizing_ops.rst:
```diff
@@ -0,0 +1,356 @@
+.. _resizing_ops:
+
+##################
+Resizing Operators
```
Much of the text here is about boundary behaviors. We should probably write a separate doc on those (since this should match) and refer it from here. That can be done later.

adler-j commented on odl/util/numerics.py:
```diff
@@ -297,6 +303,593 @@ def fast_1d_tensor_mult(ndarr, onedim_arrs, axes=None, out=None):
     return out
 
 
+def _compute_slices(arr1, arr2, num_left, pad_mode, direction):
```
could use a docstring

adler-j commented on odl/util/numerics.py:
```diff
@@ -24,8 +24,14 @@
+__all__ = ('apply_on_boundary', 'fast_1d_tensor_mult', 'resize_array')
+
+
+_SUPPORTED_PAD_MODES = ('constant', 'symmetric', 'periodic',
```
Not future proof. Change to something like `_SUPPORTED_PAD_MODES_RESIZE`

kohr-h replied:
ok

adler-j commented on odl/util/numerics.py:
```diff
@@ -297,6 +303,593 @@ def fast_1d_tensor_mult(ndarr, onedim_arrs, axes=None, out=None):
+        nearest integer. The default ``None`` is equivalent to 0.5.
+        This option can be specified per axis with a sequence.
+        Cannot be combined with ``num_left``.
+    num_left : sequence of int, optional
```
What does left even mean? Also no matching "right"?

kohr-h replied:
`num_right` can be caluclated from the shapes and `num_left`, but sure, I can add it for convenience. `left` is something I chose since it's intuitively clear in 1D, but it's less so in higher dimensions. Alternatives?
- left/right
- before/after 
- top/bottom
- ?

kohr-h replied:
How about `offset`? That should be perfectly clear and does not need a "right" counterpart. However, in that case, the default should probably be 0 in each axis. More reliable, too.

adler-j commented on odl/util/numerics.py:
```diff
@@ -297,6 +303,593 @@ def fast_1d_tensor_mult(ndarr, onedim_arrs, axes=None, out=None):
+
+    out : `numpy.ndarray`, optional
+        Array to write the result to. Must have shape ``newshp`` and
+        be able to hold the data type of the input array.
```
Data order restrictions?

kohr-h replied:
None needed? We assign stuff by slicing, that will work irrespective of the data order. Performance will depend on it, sure.

adler-j commented on odl/util/numerics.py:
```diff
@@ -297,6 +303,593 @@ def fast_1d_tensor_mult(ndarr, onedim_arrs, axes=None, out=None):
+    return out
+
+
+def _slice_dict(axis, ax_len, ndim, slc_l, slc_r):
```
Very weird way of making a function. Why not simply make it take the padding variant as an argument and return the actual slice?

kohr-h replied:
Will be refactored.

adler-j commented on odl/util/numerics.py:
```diff
@@ -297,6 +303,593 @@ def fast_1d_tensor_mult(ndarr, onedim_arrs, axes=None, out=None):
+                             '({} != {})'.format(arr.ndim, out.ndim))
+    else:
+        arr = np.asarray(arr)
+        out = np.empty(newshp, dtype=arr.dtype)
```
how about order?

kohr-h replied:
I'll add that here, yes.

kohr-h replied:
done

adler-j commented on odl/util/numerics.py:
```diff
@@ -297,6 +303,593 @@ def fast_1d_tensor_mult(ndarr, onedim_arrs, axes=None, out=None):
+            istart = num_l
+            istop = n_new - (n_add - istart)
+
+            n_pad_l = len(range(istart))
```
Wut? Why not simply `n_pad_l = istart`? Or if pad can be negative, `n_pad_l = max(0, istart)`

adler-j commented on odl/util/numerics.py:
```diff
@@ -297,6 +303,593 @@ def fast_1d_tensor_mult(ndarr, onedim_arrs, axes=None, out=None):
+                                 'order 0 padding, got 0'
+                                 ''.format(i))
+
+            if pad_mode == 'order1' and n_orig == 1:
```
`and n_orig < 2` is probably more consistent with the bellow, and also handles `n_orig = 0`.

adler-j commented on test/discr/discr_ops_test.py:
```diff
@@ -0,0 +1,181 @@
+    assert np.array_equal(out, true_res)
+
+
+def test_resizing_op_deriv():
```
This should have a fixture, and I.m.o should test much more corner cases (axis size 0 and 1, different dtypes, data orders, what happens if space is over integers? etc)

adler-j commented on test/discr/discr_ops_test.py:
```diff
@@ -0,0 +1,181 @@
+    assert res_op.is_linear
+
+
+def test_resizing_op_call(fn_impl):
```
This should have a fixture, and I.m.o should test much more corner cases (axis size 0 and 1, different dtypes, data orders, what happens if space is over integers? etc)

adler-j commented on test/discr/discr_ops_test.py:
```diff
@@ -0,0 +1,181 @@
+                         discr_kwargs={'nodes_on_bdry': True})
+
+
+def test_resizing_op_raise():
```
What about different dimensions of domain and range?

kohr-h commented on odl/util/numerics.py:
```diff
@@ -297,6 +303,593 @@ def fast_1d_tensor_mult(ndarr, onedim_arrs, axes=None, out=None):
+        raise ValueError("`pad_mode` '{}' not understood".format(pad_mode_in))
+
+    if (pad_mode == 'constant' and
+        not np.can_cast(pad_const, out.dtype) and
```
Should be `or`, right.

chongchenmath commented:
Hi, I want to use the temporary `ZeroPaddingOperator` to implement the LDDMM solver. I ever used that guy in `Linearized_deformations` branch. But I cannot find it now. So need help. Thanks a lot.

adler-j commented:
`Needs rebase`? :)

adler-j commented on doc/source/math/resizing_ops.rst:
```diff
@@ -0,0 +1,319 @@
+Introduction
+============
+In ODL, resizing of a discretized function is understood as the operation of shrinking or enlarging its domain in such a way that the size of the partition cells do not change.
+This "constant cell size" restriction is intentional since it ensures that the underlying operation can be implemented as array resizing without resampling, thus keeping those two functionalities separate.
```
Link to resampling operators/doc?

kohr-h replied:
Yes.

adler-j commented on doc/source/math/resizing_ops.rst:
```diff
@@ -0,0 +1,319 @@
+
+Basic setting
+=============
+Let now :math:`\mathbb{R}^n` with :math:`n \in \mathbb{N}` be the space of one-dimensional real vectors encoding values of a function defined on an interval :math:`[a, b] \subset \mathbb{R}`.
```
We have a long text on discretizations, how about linking to it as well?

kohr-h replied:
Well it's not really long and rather incomplete, not sure it helps much. But it can be linked anyway.

kohr-h commented:
So now I refactored the code to a great deal so it has maximum code reuse between forward and adjoint of the padding (actually resizing) operation. Tests are parametrized and check more corner cases, adjoint operator is implemented and has a test. So apart from perhaps a missing doctest for the operator, this is ready for review (and rebased :-) ).

adler-j commented on doc/source/math/resizing_ops.rst:
```diff
@@ -0,0 +1,319 @@
+
+Restriction operator
+====================
+We consider the space :math:`\mathbb{R}^m` for an :math:`m < n \in \mathbb{N}` and define the restriction operator
```
This could perhaps be done while allowing `m = n`, this would allow some useful cases such as the trivial restriction.

kohr-h replied:
It's allowed in the implementation, sure. Here, for the purpose of illustration, it would not be so interesting.

adler-j commented on doc/source/math/resizing_ops.rst:
```diff
@@ -0,0 +1,319 @@
+    c      , & \text{else}.
+    \end{cases}
+
+Note that this operator is not linear, and its derivative is the zero operator, hence the derivative of the constant padding operator is :math:`E_{\mathrm{c}}' = E_{\mathrm{z}}`.
```
the zero *padding* operator

adler-j commented on doc/source/math/resizing_ops.rst:
```diff
@@ -0,0 +1,319 @@
+Periodic padding
+----------------
+This padding mode continues the original vector :math:`x` periodically in both directions.
+For reasons of practicability, at most one whole copy is allowed on both sides, which means that the numbers :math:`n`, :math:`m` and :math:`p` need to fulfill :math:`p \leq n` ("left" padding amount) and :math:`m - (p + n) \leq n` ("right" padding amount).
```
Feels rather arbitrary. Why is this done?

kohr-h replied:
Relaxing this is certainly possible, but it would be a significant extra complexity. Instead of one slice left and right per axis, one would have to use lists, which I don't really feel like thinking about now. This is already complex enough, and if we really need multiple periods at some point, we can add it. Already now, you can get a poor man's version of that by simply extending in several steps.

adler-j commented on doc/source/math/resizing_ops.rst:
```diff
@@ -0,0 +1,319 @@
+    \end{cases}
+
+with the convention of summing values for overlapping cases, i.e. if :math:`i \in \{1, 2\}`.
+In practice, the adjoint for the order 1 padding case is applied by computing the zero'th and first order moments of :math:`y` and adding them to the two outmost entries of :math:`x` according to the above rule.
```
It is not clear for me how n-d data is handled with this interpolation.

kohr-h replied:
It's like unfolding (extension, forward) and folding (restriction, adjoint) a piece of paper. The contributions are summed. I don't know how that plays along with interpolation schemes, but they're not really involved here.

adler-j replied:
Could perhaps use some more text on this then, this was not obvious to me.

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -101,56 +111,353 @@ def _call(self, x, out=None):
+        """Constant used by this operator in case of constant padding."""
+        return self.__pad_const
+
+    @staticmethod
```
Why make this static instead of a free method?

kohr-h replied:
If it's useful on its own right, it can be a free function, sure. Here it is a bit more safe to make assumptions on the parameters.

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -101,56 +111,353 @@ def _call(self, x, out=None):
+
+class ResizingOperator(ResizingOperatorBase):
+
+    """Operator mapping a discrete function to a new domain.
```
discretized perhaps?

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -101,56 +111,353 @@ def _call(self, x, out=None):
+
+    """Operator mapping a discrete function to a new domain.
+
+    This operator is a mapping between uniformly discretized spaces with
```
Should mention `\`DiscreteLp\`` somewhere here, so users get a link to it.

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -101,56 +111,353 @@ def _call(self, x, out=None):
+        if not self.is_linear:
+            raise NotImplementedError('this operator is not linear and '
+                                      'thus has no adjoint')
+        return ResizingOperatorAdjoint(domain=self.range, range=self.domain,
```
Missing arguments like `offset` etc

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -101,56 +111,353 @@ def _call(self, x, out=None):
+        acts as left inverse, while in restriction axes, it is a
+        right inverse.
+        """
+        return ResizingOperator(domain=self.range, range=self.domain,
```
Missing arguments

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -101,56 +111,353 @@ def _call(self, x, out=None):
+    @property
+    def adjoint(self):
+        """Adjoint of this operator."""
+        return ResizingOperator(domain=self.range, range=self.domain,
```
`if not self.is_linear`

and missing arguments

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -101,56 +111,353 @@ def _call(self, x, out=None):
+
+
+def _resize_discr(discr, newshp, offset, discr_kwargs):
+    """Return ``discr`` resized to ``newshp``.
```
I disagree with this docstring, it barely retents `discr`'s properties at all, basically only taking the underlying domain and size. Things like `dspace`, `dtype` etc seem to be taken from `discr_kwargs`.

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -101,56 +111,353 @@ def _call(self, x, out=None):
+            raise ValueError('cannot combine `range` with `ran_shape`')
+
+        pad_mode = kwargs.pop('pad_mode', 'constant')
+        self.__pad_mode = str(pad_mode).lower()
```
do the usual `pad_mode_in` thingie

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -101,56 +111,353 @@ def _call(self, x, out=None):
+    See `the online documentation
+    <https://odl.readthedocs.io/math/resizing_ops.html>`_
+    on resizing operators for mathematical details.
+    """
```
This class is totally missing any examples, would be great with some simple examples.

kohr-h replied:
I'm aware that there are no doctests yet.

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -101,56 +111,353 @@ def _call(self, x, out=None):
+                                pad_const=self.pad_const)
+
+
+class ResizingOperatorAdjoint(ResizingOperatorBase):
```
Do we really need a separate operator for the `Adjoint`? It basically seems to be doing very little. Couldnt all three classes here be merged into a single class, or am I missing something?

kohr-h replied:
It doesn't do that much, no, it catches `pad_const != 0` and has a different implementation of `_call` and `adjoint`. And no inverse, which it perhaps also should have.

adler-j commented on test/util/numerics_test.py:
```diff
@@ -216,5 +222,299 @@ def test_fast_1d_tensor_mult_error():
+    resized_adj = resize_array(other_arr, array.shape, offset, pad_mode,
+                               pad_const, direction='adjoint')
+
+    print(resized)
```
leftover?

kohr-h replied:
obviously

adler-j commented:
Quite good looking pull request, had some minor points that could use fixing.

kohr-h commented:
All done except doctests.

kohr-h commented:
Added a doctest for `ResizingOp` and a use case for padding in the Fourier transform example. Also spotted some errors, fixed now. Should be ready for merge.

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -101,56 +111,407 @@ def _call(self, x, out=None):
+            For the default ``None``, a space with the same attributes
+            as ``domain`` is used, except for its shape, which is set
+            to ``ran_shp``.
+        ran_shp : sequence of int
```
, optional

kohr-h replied:
right

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -101,56 +111,407 @@ def _call(self, x, out=None):
+        providing ``ran_shp`` and, optionally, parameters for the padding
+        variant to be used. The range is inferred from ``domain`` and
+        the supplied parameters. If no ``offset`` is given, the difference
+        in size is evenly distributed to both sides:
```
What happens if the difference is a odd number?

kohr-h replied:
One more to the left, if that's an okay default.

adler-j replied:
Ok by me, but this should be mentioned.

kohr-h replied:
[It is](https://github.com/odlgroup/odl/pull/499/files/b67607feef6f368f8ccac83f2b1eb6335e5137e8#diff-783324f48a9505cd5363ed78952719f9R189)

adler-j commented on odl/discr/discr_ops.py:
```diff
@@ -101,56 +111,407 @@ def _call(self, x, out=None):
+    return tuple(offset)
+
+
+def _resize_discr(discr, newshp, offset, discr_kwargs):
```
This does not preserve `impl` and others.

kohr-h replied:
Ah f**k, forgot that. Will add a test for that, too.

adler-j commented on test/discr/discr_ops_test.py:
```diff
@@ -0,0 +1,218 @@
+# --- ResizingOperator --- #
+
+
+def test_resizing_op_init():
```
Add the `fn_impl` fixture and add `impl=fn_impl` to the spaces.

You may also want to test going from one impl to another, it is not clear if this is supported or not.

adler-j replied:
Also may want to consider storage order.

adler-j commented on test/discr/discr_ops_test.py:
```diff
@@ -0,0 +1,218 @@
+    dspace = odl.rn(3)
+    space = odl.DiscreteLp(fspace, part, dspace)
+    with pytest.raises(ValueError):
+        odl.ResizingOperator(space, ran_shp=(10,))
```
why should this raise?

kohr-h replied:
Not uniformly discretized.

adler-j commented on test/discr/discr_ops_test.py:
```diff
@@ -0,0 +1,218 @@
+    return request.param
+
+
+def test_resizing_op_properties(dtype):
```
Should also use `fn_impl` to test various fn, and use `odl.FN_IMPLS[fn_impl].available_dtypes()` to find dtypes.

adler-j replied:
Why not use the `pad_mode` fixture that you define below with this test and use it for an exhaustive test.

adler-j commented on test/discr/discr_ops_test.py:
```diff
@@ -0,0 +1,218 @@
+
+    # Different padding mode
+    res_op = odl.ResizingOperator(space, res_space, pad_const=1)
+    assert res_op.pad_const == 1.0
```
Also `assert res_op.pad_mode = 'constant'`

adler-j commented on test/discr/discr_ops_test.py:
```diff
@@ -0,0 +1,218 @@
+
+def test_resizing_op_deriv():
+
+    # Only non-trivial case is constant padding with const != 0
```
Could also do a test that `op.derivative(x) is op` for other cases?

adler-j commented on test/discr/discr_ops_test.py:
```diff
@@ -0,0 +1,218 @@
+                                  pad_const=pad_const)
+
+    # Only left inverse if the operator extentds in all axes
+    x = space.element(np.arange(space.size).reshape(space.shape))
```
Should simply use `odl.util.testutils.example_element(space)`, it would be more obvious what you mean.

adler-j commented on test/discr/discr_ops_test.py:
```diff
@@ -0,0 +1,218 @@
+
+def test_resizing_op_inverse(pad_mode, dtype, fn_impl):
+
+    if isinstance(pad_mode, tuple):
```
Instead of repeating this, why not simply add this to the fixture definition?

adler-j commented on test/util/numerics_test.py:
```diff
@@ -216,5 +222,291 @@ def test_fast_1d_tensor_mult_error():
+
+
+variants = ['extend', 'restrict', 'mixed']
+setup_params = list(product(pad_mode_params, variants))
```
It would be more clear what you meant if you instead passed the two fixtures as arguments to `resize_setup`. pytest supports "recursive" fixtures.

adler-j commented:
Some remarks on the tests, otherwise fine!

adler-j commented on test/util/numerics_test.py:
```diff
@@ -234,65 +234,73 @@ def dtype(request):
 
-    pad_mode, variant = request.param
-    pad_const = 0
+    mode, const = pad_mode
```
probably should rename these to fit the parameters in the function

`pad_mode, pad_const = padding`

or something like that.

kohr-h commented:
All test comments processed, should be mergeable.

adler-j commented on test/util/numerics_test.py:
```diff
@@ -21,13 +21,16 @@
 
-# External
-import pytest
+from itertools import product
```
unused

adler-j commented:
Should be cool, add release notes and go.
